Revise "mode" description in ansible.builtin.copy

thedoc31 opened the pull request:
Clarify the "mode" description. 

##### SUMMARY
Current wording can be misinterpreted and does not strongly or clearly enough emphasize why one would want to use both the leading zero and quote, even though just a leading zero is sufficient and will work. Ansible-lint has a decent description of this too (risky-octal rule), but it does not currently flag an unquoted number with a leading zero, so we can't say in this documentation that it's required to do both.

See further conversation in #79991

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
ansible.builtin.copy

##### ADDITIONAL INFORMATION
If accepted, this should also be updated in the documentation for
ansible.builtin.assemble
ansible.builtin.file
ansible.builtin.replace
ansible.builtin.template
And all other examples in the documentation of a numeric mode should be updated for consistency following the recommendations above.

thedoc31 committed:
Update copy.py

Clarify the "mode" description. Current wording can be misinterpreted and does not strongly or clearly enough emphasize why one would want to use both the leading zero and quote, even though just a leading zero is sufficient and will work. Ansible-lint has a decent description of this too (risky-octal rule), but it does not currently flag an unquoted number with a leading zero, so we can't say in this documentation that it's required to do both.

thedoc31 committed:
Remove trailing whitespace

Removing trailing whitespace to resolve linting errors

oraNod commented on lib/ansible/modules/copy.py:
```diff
@@ -63,11 +63,11 @@
-      one of these rules will end up with a decimal number which will have unexpected results.
+    - For those used to C(/usr/bin/chmod), remember that modes are actually octal numbers.
+      To prevent unexpected results of decimal to octal conversion, you B(must) add a leading zero
+      (like C(0644) or C(01777)) to the number. Though not required, you should also quote the number
+      (like C('0644') or C('01777')) to make it a string. The combination of these two elements guarantee
+      Ansible receives a string the YAML parser can interpret correctly into the desired permissions.
```
```suggestion
      (like C(0644) or C(01777)) to the number. You can also wrap the number in single quotes
      (like C('0644') or C('01777')) to pass it to the YAML parser as a string. This guarantees that 
      Ansible correctly interprets the desired permissions.
```

oraNod replied:
@thedoc31 Nice catch here and thanks for the contribution. Feel free to take or leave my suggestion.

thedoc31 committed:
Update lib/ansible/modules/copy.py

Co-authored-by: Don Naro <dnaro@redhat.com>

thedoc31 committed:
Remove pesky trailing space

samccann commented:
https://github.com/ansible/ansible/pull/80112 seems to say something different (different modules but similar original text).

thedoc31 committed:
Update copy.py

#78715 and #80112 explain another situation where even a leading zero alone still caused unintended conversion issues. Therefore, the guidance should be to use both at all times instead of making the string quoting optional.

thedoc31 commented:
> #80112 seems to say something different (different modules but similar original text).

Very interesting, @samccann. I revised this once more based on the very technical discussion in #78715. That has enough detail which leads me to believe an unquoted decimal should _never_ be used in any circumstance - it should always be a string. I also commented in #80112 to suggest the same change be made there with the same wording. If these are approved by the community I'll submit PRs for the other modules which have a "mode" description in them.

felixfontein commented on lib/ansible/modules/copy.py:
```diff
@@ -63,11 +63,10 @@
-      one of these rules will end up with a decimal number which will have unexpected results.
+    - For those used to C(/usr/bin/chmod), remember that modes are actually octal numbers.
+      To prevent unexpected results of decimal to octal conversion, you B(must) add a leading zero
+      to the number (B)and wrap the number in quotes (like C('0644') or C('01777')) to pass it to
```
The `(B)` in this line is wrong.

thedoc31 committed:
Fix incorrect markdown

Fixes incorrect markdown usage

samccann commented on lib/ansible/modules/copy.py:
```diff
@@ -63,11 +63,10 @@
+    - For those used to C(/usr/bin/chmod), remember that modes are actually octal numbers.
+      To prevent unexpected results of decimal to octal conversion, you B(must) add a leading zero
+      to the number B(and) wrap the number in quotes (like C('0644') or C('01777')) to pass it to
+      the YAML parser as a string. This guarantees that Ansible correctly interprets the desired permissions.
```
Sorry for the delay here. I'm trying to catch up on all the discussions and differing PRs.

A couple of comments:
1 - Ansible tends to stay clear of 'must' statements. As I read the different discussions, octals 'can' work. In that sense, what is documented in https://github.com/ansible/ansible/pull/80112/files is clearer on this point so I'm inclined to use that PR as the model for all of these 'mode'. While your description matches what ansible-lint recommends,  that is for the higher linting level so to speak (security).  And historically, ansible-core leaves in all the flexibility in its descriptions (as in very few recommended practices, more like notes to warn you were things could go wrong, like that other PR).

samccann commented:
Closing in favor of https://github.com/ansible/ansible/pull/80112 but thanks for all the help in understanding the intricacies of this @thedoc31 !
